File: My First App/My First App/ViewController.m
//
//  ViewController.m
//
//
//
//  Copyright (c) 2016 Magnet Systems, Inc. All rights reserved.
//

#import "ViewController.h"

@import MagnetMax;


@interface ViewController ()

@property (nonatomic, weak) IBOutlet UITextField *messageText;

@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Register for a notification to receive the message
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didReceiveMessage:)
                                                 name:MMXDidReceiveMessageNotification
                                               object:nil];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // Begin the user registration process
    [self createAndRegisterUser];
}


/***
 *
 *
 *                                      MAGNET MESSAGE DEMO METHODS
 *
 *   (This view controller is preconfigured with settings and code to send and recieve basic messages)
 *                         See documentation at https://developer.magnet.com/docs
 *
 *
 *
 ***/


//MARK: Notifications


- (void)didReceiveMessage:(NSNotification *)notification {
    NSDictionary *userInfo = notification.userInfo;
    //retrieve MMXMessage from userInfo
    MMXMessage *message = userInfo[MMXMessageKey];
    
    [self showAlert:@"Received message"
            message:[NSString stringWithFormat:@"content: %@", message.messageContent]];
    NSLog(@"Received message content: %@", message.messageContent);
}


//MARK: Private Methods


// Creates and registers a generic user.
- (void)createAndRegisterUser {
    // Create a user object
    MMUser *user = [[MMUser alloc] init];
    
    // Optional
    user.firstName = @"John";
    user.lastName = @"Doe";
    
    // Required
    user.userName = @"john.doe@magnet.com";
    user.password = @"magnet";
    
    // Register the user
    [user register:^(MMUser * currentUser) {
        [self loginUser:currentUser.userName password:currentUser.password];
    } failure:^(NSError * error) {
        // 409 = the user is already registered
        if (error.code == 409) {
            // If a user is already already registered automatically login
            [self loginUser:user.userName password:user.password];
        }else {
            NSLog(@"[ERROR]: Failed to register: %@", error.localizedDescription);
        }
    }];
}

- (void)loginUser:(NSString *)username password:(NSString *)password {
    // Log in user with an NSURLCredential
    NSURLCredential *credential = [[NSURLCredential alloc] initWithUser:username password:password persistence:NSURLCredentialPersistenceNone];
    [MMUser login:credential success:^{
        
        // Indicate that you are ready to receive messages now!
        [MMX start];
        
        // Send an initial message
        [self sendMessage:self.messageText.text];
    } failure:^(NSError * error) {
        NSLog(@"[ERROR]: Failed to login: %@", error.localizedDescription);
    }];
}

- (void)sendMessage:(NSString *)messageString {
    if (![MMUser currentUser] || !messageString) {
        return;
    }
    // Dictionary to send
    NSDictionary *content = @{@"message": messageString};
    // Create message recipient -> yourself
    MMXMessage *message = [MMXMessage messageToRecipients:[NSSet setWithObject:[MMUser currentUser]] messageContent:content];
    
    // Uncomment to add attachment
    //    MMAttachment *attachment = [[MMAttachment alloc]  initWithFileURL:imageUrl mimeType:@"image/jpg"];
    //    [message addAttachment:attachment];
    
    //send the message
    [message sendWithSuccess:^(NSSet<NSString *> *invalidUsers) {
        NSLog(@"[INFO]: Sent message with content: %@", message.messageContent);
    } failure:^(NSError *error) {
        NSLog(@"[ERROR]: Failed to send message: %@", error.localizedDescription);
    }];
}

- (void)showAlert:(NSString *)title message:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                             message:message
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"ok"
                                                        style:UIAlertActionStyleDefault
                                                      handler:^(UIAlertAction * action) {
                                                          [alertController dismissViewControllerAnimated:NO completion:nil];
                                                      }]];
    
    [self presentViewController:alertController animated:NO completion:nil];
}


//MARK: Actions


- (IBAction)send:(id)sender {
    [self sendMessage:self.messageText.text];
}

@end

File: QuickStart/QuickStart/ViewController.swift
/*
* Copyright (c) 2015 Magnet Systems, Inc.
* All rights reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License"); you
* may not use this file except in compliance with the License. You
* may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
* implied. See the License for the specific language governing
* permissions and limitations under the License.
*/

import UIKit
import MagnetMax

class ViewController: UIViewController {
    
    
    // MARK: Outlets
    
    
    @IBOutlet weak var passwordTextField: UITextField!
    @IBOutlet weak var userNameTextField: UITextField!
    
    
    // MARK: Overrides
    
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        MMUser.logout({ () -> Void in
            self.userNameTextField.text = nil
            self.passwordTextField.text = nil
            }) { (error) -> Void in
                print("[ERROR]: \(error)")
        }
    }
    
    
    // MARK: Private implementation
    
    
    private enum InputError: ErrorType {
        case InvalidUserName
        case InvalidPassword
    }
    
    private func validateCredential() throws -> (String, String) {
        // Get values from UI
        guard let userName = userNameTextField.text where (userName.isEmpty == false) else {
            throw InputError.InvalidUserName
        }
        
        guard let password = passwordTextField.text where (password.isEmpty == false) else {
            throw InputError.InvalidPassword
        }
        
        return (userName, password)
    }
    
    
    // MARK: Actions
    
    
    @IBAction func login() {
        do {
            // Validate
            let (userName, password) = try validateCredential()
            
            // Login
            let credential = NSURLCredential(user: userName, password: password, persistence: .None)
            
            MMUser.login(credential,
                success: {
                    print(MMUser.currentUser())
                    self.performSegueWithIdentifier("featuresSegue", sender: nil)
                },
                failure: { error in
                    print("[ERROR]: \(error.localizedDescription)")
            })
        } catch {
            print("[ERROR]: Could not login because of invalid userName or password")
        }
    }
    
    @IBAction func redirectToLoginScreenSegue(segue: UIStoryboardSegue) {
        print("[INFO]: User was logged out, redirected from Features screen")
    }
    
    @IBAction func register() {
        do {
            // Validate
            let (userName, password) = try validateCredential()
            
            // Register
            let user = MMUser()
            user.userName = userName
            user.firstName = userName
            user.password = password
            
            user.register({ [weak self] user in
                self?.login()
                }, failure: { [weak self] error in
                    if error.code == 409 {
                        // The user already exists, let's attempt a login
                        self?.login()
                    }
                    print("[ERROR]: \(error)")
                })
        } catch {
            print("[ERROR]: Could not register because of invalid userName or password")
        }
    }
    
}

